File: src/com/rian/difficultycalculator/skills/Aim.java
package com.rian.difficultycalculator.skills;

import com.rian.difficultycalculator.beatmap.hitobject.DifficultyHitObject;
import com.rian.difficultycalculator.evaluators.AimEvaluator;

import java.util.EnumSet;

import ru.nsu.ccfit.zuev.osu.game.mods.GameMod;


/**
 * Represents the skill required to correctly aim at every object in the map with a uniform circle size and normalized distances.
 */
public class Aim extends StrainSkill {
    private final boolean withSliders;
    private double currentStrain;

    /**
     * @param mods The mods that this skill processes.
     * @param withSliders Whether to consider sliders in the calculation.
     */
    public Aim(EnumSet<GameMod> mods, boolean withSliders) {
        super(mods);

        this.withSliders = withSliders;
    }

    @Override
    protected double strainValueAt(DifficultyHitObject current) {
        currentStrain *= strainDecay(current.deltaTime);
        double skillMultiplier = 23.55;
        currentStrain += AimEvaluator.evaluateDifficultyOf(current, withSliders) * skillMultiplier;

        return currentStrain;
    }

    @Override
    protected double calculateInitialStrain(double time, DifficultyHitObject current) {
        return currentStrain * strainDecay(time - current.previous(0).startTime);
    }

    @Override
    protected void saveToHitObject(DifficultyHitObject current) {
        if (withSliders) {
            current.aimStrainWithSliders = currentStrain;
        } else {
            current.aimStrainWithoutSliders = currentStrain;
        }
    }

    private double strainDecay(double ms) {
        return Math.pow(0.15, ms / 1000);
    }
}

File: src/com/rian/osu/difficulty/skills/Aim.kt
package com.rian.osu.difficulty.skills

import com.rian.osu.difficulty.DifficultyHitObject
import com.rian.osu.difficulty.evaluators.AimEvaluator.evaluateDifficultyOf
import com.rian.osu.mods.Mod
import kotlin.math.pow

/**
 * Represents the skill required to correctly aim at every object in the map with a uniform circle size and normalized distances.
 */
class Aim(
    /**
     * The [Mod]s that this skill processes.
     */
    mods: List<Mod>,

    /**
     * Whether to consider sliders in the calculation.
     */
    private val withSliders: Boolean
) : StrainSkill(mods) {
    private var currentStrain = 0.0
    private val skillMultiplier = 23.55
    private val strainDecayBase = 0.15

    override fun strainValueAt(current: DifficultyHitObject): Double {
        currentStrain *= strainDecay(current.deltaTime)
        currentStrain += evaluateDifficultyOf(current, withSliders) * skillMultiplier

        return currentStrain
    }

    override fun calculateInitialStrain(time: Double, current: DifficultyHitObject) =
        currentStrain * strainDecay(time - current.previous(0)!!.startTime)

    override fun saveToHitObject(current: DifficultyHitObject) {
        if (withSliders) {
            current.aimStrainWithSliders = currentStrain
        } else {
            current.aimStrainWithoutSliders = currentStrain
        }
    }

    private fun strainDecay(ms: Double) = strainDecayBase.pow(ms / 1000)
}
